pkg/loadbalancer/experimental/healthserver.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+
+	cmtypes "github.com/cilium/cilium/pkg/clustermesh/types"
+	lb "github.com/cilium/cilium/pkg/loadbalancer"
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/rate"
+	"github.com/cilium/cilium/pkg/source"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// healthServerCell implements the support for the HealthCheckNodePort field.
+// For each service that has HealthCheckNodePort set to a non-zero value it
+// starts an HTTP server on that port and responds with the number of active
+// backends. This allows external load-balancers to avoid directing traffic
+// to services that have no available backends.
+var healthServerCell = cell.Module(
+	"healthserver",
+	"Serves NodePort health status to external load-balancers",
+	cell.Invoke(registerHealthServer),
+)
+
+type healthServerParams struct {
+	cell.In
+
+	Jobs      job.Group
+	Log       *slog.Logger
+	DB        *statedb.DB
+	ExtConfig ExternalConfig
+	Frontends statedb.Table[*Frontend]
+	Backends  statedb.Table[*Backend]
+	Writer    *Writer
+}
+
+// healthServer manages HTTP health check ports. For each added service,
+// it opens a HTTP server on the specified HealthCheckNodePort and either
+// responds with 200 OK if there are local endpoints for the service, or with
+// 503 Service Unavailable if the service does not have any local endpoints.
+type healthServer struct {
+	params        healthServerParams
+	serverByPort  map[uint16]*httpHealthServer
+	portByService map[lb.ServiceName]uint16
+	nodeName      string
+}
+
+func registerHealthServer(params healthServerParams) {
+	if !params.ExtConfig.EnableHealthCheckNodePort {
+		return
+	}
+
+	s := &healthServer{
+		params:        params,
+		serverByPort:  map[uint16]*httpHealthServer{},
+		portByService: map[lb.ServiceName]uint16{},
+	}
+	params.Jobs.Add(job.OneShot("control-loop", s.controlLoop))
+}
+
+var (
+	// healthServerAddr is the IP address on which the health HTTP servers
+	// listen on.
+	healthServerAddr = cmtypes.MustParseAddrCluster("127.0.0.88")
+)
+
+func (s *healthServer) controlLoop(ctx context.Context, health cell.Health) error {
+	s.nodeName = nodeTypes.GetName()
+
+	// Watch services for changes to add and remove the listeners.
+	wtxn := s.params.DB.WriteTxn(s.params.Frontends)
+	frontendChanges, err := s.params.Frontends.Changes(wtxn)
+	wtxn.Commit()
+	if err != nil {
+		return err
+	}
+
+	// Limit the rate at which the change batches are processed.
+	limiter := rate.NewLimiter(100*time.Millisecond, 1)
+
+	defer s.cleanupListeners(ctx)
+
+	for {
+		if err := limiter.Wait(ctx); err != nil {
+			return err
+		}
+
+		changes, watch := frontendChanges.Next(s.params.DB.ReadTxn())
+		for change := range changes {
+			fe := change.Object
+			svc := fe.service
+			name := svc.Name
+			healthServiceName := name.AppendSuffix("-healthserver")
+			port := svc.HealthCheckNodePort
+
+			if !isExtLocal(fe) {
+				continue
+			}
+
```

Comment:
Old `upsertService()` code seems to support that a service was marked `eTP:local` in the past, but then gets modified. I suppose that's a bit too crazy to continue supporting?